lib/spack/spack/database.py

```diff
@@ -695,8 +695,32 @@ def _read_spec_from_dict(self, spec_reader, hash_key, installs, hash=ht.dag_hash
 
         # Build spec from dict first.
         spec = spec_reader.from_node_dict(spec_dict)
+
+        self._validate_version(spec)
+
         return spec
 
+    def _validate_version(self, spec):
```

Comment:
So, I like this fix, but I think it needs to apply in more places.  If we read a concrete spec from a lock file, a build cache, or an upstream, what happens?  Upstreams and build caches are handled here, I guess, but lockfiles?  Random `spec.json` files? (this last one I care less about)

Reply:
FWIW, this can be handled much of the time in Spec, the class. That applies in the vast majority of places, I *think* this is a workaround for one of the few places that isn't enough.

Reply:
I've moved it into `_mark_root_concrete`; it should work for specs from the DB, lock files, and spec.json files.